_overviews/hello-scala/for-expressions.md

```diff
@@ -0,0 +1,144 @@
+```
+
+That expression can be read as, “For every number `n` in the list of numbers `nums`, double each value, and then assign all of the new values to the variable `doubledNums`.” This is what it looks like in the Scala REPL:
+
+```scala
+scala> val doubledNums = for (n <- nums) yield n * 2
+doubledNums: Seq[Int] = List(2, 4, 6)
+```
+
+As the REPL output shows, the new list `doubledNums` contains these values:
+
+```scala
+List(2,4,6)
+```
+
+In summary, the result of the for-expression is that it creates a new variable named `doubledNums` whose values were created by doubling each value in the original list, `nums`.
+
+
+
+## Capitalizing a list of strings
+
+You can use the same approach with a list of strings. For example, given this list of lowercase strings:
+
+```scala
+val names = List("adam", "david", "frank")
+```
+
+You can create a list of capitalized strings with this for-expression:
+
+```scala
+val ucNames = for (name <- names) yield name.capitalize
+```
+
+The REPL shows how this works:
+
+```scala
+scala> val ucNames = for (name <- names) yield name.capitalize
+ucNames: List[String] = List(Adam, David, Frank)
+```
+
+Success! Each name in the new variable `ucNames` is capitalized.
+
+
+
+## The `yield` keyword
+
+Notice that both of those for-expressions use the `yield` keyword:
+
+```scala
+val doubledNums = for (n <- nums) yield n * 2
+                                  -----
+
+val ucNames = for (name <- names) yield name.capitalize
+                                  -----
+```
+
+Using `yield` after `for` is the “secret sauce” that says, “I want to yield a new collection from the existing collection that I’m iterating over in the for-expression, using the algorithm shown.”
+
+
+
+## Using a block of code after `yield`
+
+The code after the `yield` expression can be as long as necessary to solve the current problem. For example, given a list of strings like this:
+
+```scala
+val names = List("_adam", "_david", "_frank")
+```
+
+Imagine that you want to create a new list that has the capitalized names of each person. To do that, you first need to remove the underscore character at the beginning of each name, and then capitalize each name. To remove the underscore from each name, you call the `tail` method on each `String`. After you do that, you call the `capitalize` method on each string. Here’s how you can use a for-expression to solve this problem:
+
+```scala
+val capNames = for (name <- names) yield {
+    val nameWithoutUnderscore = name.tail
+    val capName = nameWithoutUnderscore.capitalize
+    capName
+}
+```
+
+If you put that code in the REPL, you’ll see this result:
+
+```scala
+capNames: List[String] = List(Adam, David, Frank)
+```
+
+
+### How `tail` works
+
+The `tail` method works on sequential collections, and returns every element in the collection other than the first element (which is known as the “head” element). Because a `String` is also a linear collection, `tail` works on strings like this:
```

Comment:
Agreed, at least myself I use `StringBuilder`s and `StringBuffer`s whenever I have to manipulate strings. We can either remove the section or leave it and just say that doing `tail` on strings is discouraged because of its performance penalties.

Reply:
I went ahead and removed this because we converted `tail` to `drop(1)` in the preceding discussion.